### PR TITLE
Data flow: Add `ConfigSig::accessPathLimit`

### PR DESCRIPTION
**Pull request by hvitved**
Adds an option to control the access path limit on a per-configuration basis. Needed for https://github.com/github/codeql/pull/15866.

### CODE REVIEW

**Review by atorralba — DISMISSED** [review]
As I understand it, we already had a default access path limit of 5, so this shouldn't cause any changes in results.

File: shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll [hunk 1]
```diff
@@ -1328,6 +1331,13 @@ module MakeImpl<InputSig Lang> {
           fwdFlow1(_, _, _, _, _, _, t0, t, ap, _) and t0 != t
         }
 
+        bindingset[c, t, tail]
+        additional Ap apCons(Content c, Typ t, Ap tail) {
+          result = Param::apCons(c, t, tail) and
+          Config::accessPathLimit() > 0 and
+          if tail instanceof ApNil then any() else Config::accessPathLimit() > 1
```
Comment (michaelnebel) [inline shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll:1338]:
```suggestion
          (
            tail instanceof ApNil and Config::accessPathLimit() > 0
            or
            Config::accessPathLimit() > 1
          )
```

**Comment by hvitved** [comment]
> As I understand it, we already had a default access path limit of 5, so this shouldn't cause any changes in results.

Correct, no result changes are expected.

File: shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll (hunk 1 above)
Comment (hvitved) [reply shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll]: Thanks; I did a slightly different rewrite.

File: shared/dataflow/codeql/dataflow/DataFlow.qll [hunk 2]
```diff
@@ -599,6 +607,8 @@ module DataFlowMake<InputSig Lang> {
   module GlobalWithState<StateConfigSig Config> implements GlobalFlowSig {
     private module C implements FullStateConfigSig {
       import Config
+
+      predicate accessPathLimit = Config::accessPathLimit/0;
```
Comment (aschackmull) [inline shared/dataflow/codeql/dataflow/DataFlow.qll:611]: same

File: shared/dataflow/codeql/dataflow/DataFlow.qll [hunk 3]
```diff
@@ -583,6 +589,8 @@ module DataFlowMake<InputSig Lang> {
     private module C implements FullStateConfigSig {
       import DefaultState<Config>
       import Config
+
+      predicate accessPathLimit = Config::accessPathLimit/0;
```
[the remark this answers is not in the document]
Comment (hvitved) [reply shared/dataflow/codeql/dataflow/DataFlow.qll:593]: It is needed:
```
The predicate Config::accessPathLimit(), which was brought into scope by this import, conflicts with existing declarations: predicate Lang::accessPathLimit() from DataFlow.qll:265
```

[the remark this answers is not in the document]
Comment (aschackmull) [reply shared/dataflow/codeql/dataflow/DataFlow.qll:593]: Right, the `Lang` version is in scope.

**Review by michaelnebel — DISMISSED** [review]
C# LGTM!

File: shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll [hunk 4]
```diff
@@ -1328,6 +1331,15 @@ module MakeImpl<InputSig Lang> {
           fwdFlow1(_, _, _, _, _, _, t0, t, ap, _) and t0 != t
         }
 
+        bindingset[c, t, tail]
+        additional Ap apCons(Content c, Typ t, Ap tail) {
+          result = Param::apCons(c, t, tail) and
+          exists(int limit |
+            limit = Config::accessPathLimit() and
+            if tail instanceof ApNil then limit > 0 else limit > 1
+          )
```
Comment (aschackmull) [inline shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll:1340]: How does this affect DIL/RA? Does it compile away inder normal circumstances?

File: shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll [hunk 5]
```diff
@@ -3189,7 +3201,10 @@ module MakeImpl<InputSig Lang> {
       Typ getTyp(DataFlowType t) { result = t }
 
       bindingset[c, t, tail]
-      Ap apCons(Content c, Typ t, Ap tail) { result.isCons(c, t, tail) }
+      Ap apCons(Content c, Typ t, Ap tail) {
+        result.isCons(c, t, tail) and
+        Config::accessPathLimit() > tail.len()
```
Comment (aschackmull) [inline shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll:3206]: This should not be necessary - just ensure that all `AccessPathApprox` satisfy the limit.

Comment (aschackmull) [reply shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll:3206]: Most do, you just need the `> 0` check in the `ConsNil` case, I think.

**Review by aschackmull — COMMENTED** [review]
I think you missed the reference to `accessPathLimit` in partial flow.

File: shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll (hunk 4 above)
Comment (aschackmull) [reply shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll:1340]: Actually, you can move the `limit > 0` check into the `useFieldFlow` predicate. Then you get filtering already in stage 1.

Comment (aschackmull) [reply shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll:1340]: ... after which you can assume `limit > 0`, so no need to tweak the `ConsNil` case in `AccessPathApprox` after all.

Comment (hvitved) [reply shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll:1340]: Sadly, no, but the optimizer should be able to. The added DIL is
```
        (
          (
            `DataFlowImpl::Impl<SqlInjectionQuery::SqlInjectionFlow::C>::Stage2Param::ApNil#ec757912`(ap0) and
            GT(5, 0)
          )
          or
          (
            not(
              `DataFlowImpl::Impl<SqlInjectionQuery::SqlInjectionFlow::C>::Stage2Param::ApNil#ec757912`(ap0)
            ) and
            GT(5, 1)
          )
        )

```

File: shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll (hunk 5 above)
Comment (hvitved) [reply shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll:3206]: Right.

File: shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll (hunk 4 above)
Comment (aschackmull) [reply shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll:1340]: And you really only need the `limit > 1` check in the `Stage2Param` version of `apCons` - subsequent stages will then get the condition by invariant.

**Review by owen-mc — DISMISSED** [review]
Go 👍🏻

File: shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll [hunk 6]
```diff
@@ -2522,7 +2527,11 @@ module MakeImpl<InputSig Lang> {
 
       bindingset[c, t, tail]
       Ap apCons(Content c, Typ t, Ap tail) {
-        result = true and exists(c) and exists(t) and exists(tail)
+        result = true and
+        exists(c) and
+        exists(t) and
+        exists(tail) and
```
Comment (aschackmull) [inline shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll:2533]: This line is superfluous now.
```suggestion
```

Comment (hvitved) [reply shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll]: Fixed.

**Review by aschackmull — APPROVED** [review]
LGTM now

**Pull request merged**